Don't compile rustc to self-test compiletest

jyn514 opened the pull request:
This was changed from stage 0 to 1 in https://github.com/rust-lang/rust/pull/108905, but I'm not sure why. Change it to `top_stage` instead to allow people to choose the stage.

This should save quite a bit of time in the `mingw-check` builder, which explicitly runs `x test --stage 0 compiletest`.

Note that this also fixes a latent bug that depended on running `x build compiler` before `x doc compiler`, as well as a couple cleanups related to symlinks (which made the latent bug easier to find).

cc @pietroalbini

jyn514 committed:
Don't compile rustc to self-test compiletest

This was changed from stage 0 to 1 in https://github.com/rust-lang/rust/pull/108905, but I'm not
sure why. Change it to `top_stage` instead to allow people to choose the stage.

This should save quite a bit of time in the `mingw-check` builder, which explicitly runs `x test --stage 0 compiletest`.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5165253201/jobs/9304655862) [(plain)](https://github.com/rust-lang/rust/commit/f6611d34f13be7b105a6eda384aa25e30f3c733b/checks/13980943145/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Finished release [optimized] target(s) in 0.10s
##[endgroup]
##[group]Documenting compiler (stage0 -> stage1, x86_64-unknown-linux-gnu)
Documenting compiler (stage0 -> stage1, x86_64-unknown-linux-gnu)
thread 'main' panicked at 'symlink_dir_force(&builder.config, &out, &out_dir) failed with No such file or directory (os error 2)', doc.rs:697:9
##[endgroup]
Build completed unsuccessfully in 0:00:00
```

</details>

jyn514 committed:
rename src/dst to original/link

this is consistent with std and makes it much easier to understand what's actually happening

jyn514 committed:
don't return a `Result` from symlink_dir_force

this gives a more helpful backtrace if it fails

before:
```
  thread 'main' panicked at 'symlink_dir_force(&builder.config, &out, &out_dir) failed with No such file or directory (os error 2)', doc.rs:697:9
```

after:
```
thread 'main' panicked at 'symlink_dir(config, original, link) failed with No such file or directory (os error 2) ("failed to create link from /home/jyn/src/rust2/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/doc -> /home/jyn/src/rust2/build/x86_64-unknown-linux-gnu/compiler-doc")', doc.rs:975:5
```

jyn514 committed:
fix `x doc --stage 0 compiler` if the compiler hasn't yet been built

pietroalbini commented:
...not sure why I did that, but *maybe* it was because I needed the new `--print=all-target-specs-json` during tests? What happens if you revert it to zero?

jyn514 commented:
it works fine - this is the stage compiletest is *built with*, which has no relation to the compiler under test.

Mark-Simulacrum commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit d613134623d1659ece542a02b585b70cbf36c2b5 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum d613134623d1659ece542a02b585b70cbf36c2b5 -->
<!-- homu: {"type":"Approved","sha":"d613134623d1659ece542a02b585b70cbf36c2b5","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit d613134623d1659ece542a02b585b70cbf36c2b5 with merge b8a50010de397df570b38fe67bda435b665e2d86...
<!-- homu: {"type":"BuildStarted","head_sha":"d613134623d1659ece542a02b585b70cbf36c2b5","merge_sha":"b8a50010de397df570b38fe67bda435b665e2d86"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5232077633/jobs/9447601886)
Approved by: Mark-Simulacrum
Pushing b8a50010de397df570b38fe67bda435b665e2d86 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5232077633/jobs/9447601886"},"merge_sha":"b8a50010de397df570b38fe67bda435b665e2d86"} -->

rust-timer commented:
Finished benchmarking commit (b8a50010de397df570b38fe67bda435b665e2d86): [comparison URL](https://perf.rust-lang.org/compare.html?start=7820972f866ca2cea281ccc39201c0cd27087110&end=b8a50010de397df570b38fe67bda435b665e2d86&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=7820972f866ca2cea281ccc39201c0cd27087110&end=b8a50010de397df570b38fe67bda435b665e2d86&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=7820972f866ca2cea281ccc39201c0cd27087110&end=b8a50010de397df570b38fe67bda435b665e2d86&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.1%, -2.1%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=7820972f866ca2cea281ccc39201c0cd27087110&end=b8a50010de397df570b38fe67bda435b665e2d86&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=7820972f866ca2cea281ccc39201c0cd27087110&end=b8a50010de397df570b38fe67bda435b665e2d86&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 649.872s -> 649.051s (-0.13%)
